File: Broadcasts/src/main/java/com/example/shad2018_practical4/broadcast/MainActivity.java
package com.example.shad2018_practical4.broadcast;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private BroadcastReceiver mMonitor = new BroadcastReceiver() {
        @Override
        public void onReceive(final Context context, final Intent intent) {
            String action = intent.getAction();
            String data = intent.getDataString();
            StringBuilder logMessage = new StringBuilder()
                    .append(">").append(action)
                    .append(" for package = ").append(data)
                    .append("\n\n");
            mPackageLog.append(logMessage);
            mTextView.setText(mPackageLog.toString());
            Log.i("Shad", logMessage.toString());
        }
    };

    private TextView mTextView;
    private StringBuilder mPackageLog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mTextView = findViewById(R.id.tv_main);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mPackageLog = new StringBuilder().append("Packages log: ").append("\n\n");
        mTextView.setText(mPackageLog.toString());

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(Intent.ACTION_PACKAGE_ADDED);
        intentFilter.addAction(Intent.ACTION_PACKAGE_REPLACED);
        intentFilter.addAction(Intent.ACTION_PACKAGE_REMOVED);
        intentFilter.addDataScheme("package");
        registerReceiver(mMonitor, intentFilter);
    }

    @Override
    protected void onStop() {
        super.onStop();
        unregisterReceiver(mMonitor);
    }
}

File: serializableparcelable/src/main/java/com/exmaple/shad/serializableparcelable/MainActivity.kt
package com.exmaple.shad.serializableparcelable

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button

class MainActivity : AppCompatActivity() {

    private val catKey = "com.example.shad.cat"
    private val dogKey = "com.example.shad.dog"
    private lateinit var leia: Cat
    private lateinit var ghost: Dog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.newActivityBtn).setOnClickListener {
            startNewActivity()
        }

        leia = Cat("Leia", 7, "grey")
//        ghost = Dog("Ghost", 5, "white")
        ghost = Dog("Ghost", 5, "white", leia)
    }

    private fun startNewActivity() {
        val intent = Intent(this, AnotherActivity::class.java)
        intent.putExtra(catKey, leia)
        intent.putExtra(dogKey, ghost)
        startActivity(intent)
    }
}
